File: ts/src/schema/migration.ts
// Schema migration — lazy per-row migration on read

import type { MigrationStep, StoredSchema, CompiledSchema } from "../types.ts";

interface MigrationChainStep {
  fromVersion: number;
  toVersion: number;
  rename?: Record<string, string>;
  transform?: (row: Record<string, unknown>) => Record<string, unknown>;
  addedFields: string[];
  removedFields: string[];
  targetSchema: StoredSchema;
}

export class MigrationChain {
  readonly steps: MigrationChainStep[];

  constructor(steps: MigrationChainStep[]) {
    this.steps = steps;
  }

  migrate(row: Record<string, unknown>): Record<string, unknown> {
    let current = { ...row };

    for (const step of this.steps) {
      // Apply renames
      if (step.rename) {
        for (const [oldName, newName] of Object.entries(step.rename)) {
          if (oldName in current) {
            current[newName] = current[oldName];
            delete current[oldName];
          }
        }
      }

      // Apply custom transform
      if (step.transform) {
        current = step.transform(current);
      }

      // Add new fields with null defaults
      for (const field of step.addedFields) {
        if (!(field in current)) {
          current[field] = null;
        }
      }

      // Remove old fields
      for (const field of step.removedFields) {
        delete current[field];
      }
    }

    return current;
  }
}

export function buildMigrationChain(
  fromVersion: number,
  toVersion: number,
  migrations: MigrationStep[],
  schemas: Record<number, StoredSchema>,
): MigrationChain {
  const steps: MigrationChainStep[] = [];

  for (let v = fromVersion + 1; v <= toVersion; v++) {
    const migration = migrations.find((m) => m.version === v);
    const prevSchema = schemas[v - 1];
    const targetSchema = schemas[v];

    if (!targetSchema) {
      throw new Error(`Missing schema definition for version ${v}`);
    }

    const prevFieldNames = new Set(prevSchema?.columns.map((c) => c.name) ?? []);
    const targetFieldNames = new Set(targetSchema.columns.map((c) => c.name));

    // Determine renames from migration hints
    const renameOldNames = new Set(Object.keys(migration?.rename ?? {}));
    const renameNewNames = new Set(Object.values(migration?.rename ?? {}));

    // Added: in target but not in prev, and not a rename target
    const addedFields = [...targetFieldNames].filter(
      (f) => !prevFieldNames.has(f) && !renameNewNames.has(f),
    );

    // Removed: in prev but not in target, and not a rename source
    const removedFields = [...prevFieldNames].filter(
      (f) => !targetFieldNames.has(f) && !renameOldNames.has(f),
    );

    steps.push({
      fromVersion: v - 1,
      toVersion: v,
      rename: migration?.rename,
      transform: migration?.transform,
      addedFields,
      removedFields,
      targetSchema,
    });
  }

  return new MigrationChain(steps);
}

export function deserializeWithSchema(
  fieldValues: unknown[],
  schema: StoredSchema,
): Record<string, unknown> {
  const row: Record<string, unknown> = {};
  for (let i = 0; i < schema.columns.length && i < fieldValues.length; i++) {
    row[schema.columns[i].name] = fieldValues[i];
  }
  return row;
}

File: internal/schema/migration.go
package schema

// MigrationChainStep describes one version transition.
type MigrationChainStep struct {
	FromVersion   int
	ToVersion     int
	Rename        map[string]string
	AddedFields   []string
	RemovedFields []string
	TargetSchema  *StoredSchema
}

// MigrationChain applies a series of migration steps to a row.
type MigrationChain struct {
	Steps []MigrationChainStep
}

// Migrate applies all chain steps to transform a row from old schema to current.
// Note: transform functions are handled on the JS side. Go only handles
// renames, added fields (null defaults), and removed fields.
func (mc *MigrationChain) Migrate(row map[string]interface{}) map[string]interface{} {
	current := make(map[string]interface{}, len(row))
	for k, v := range row {
		current[k] = v
	}

	for _, step := range mc.Steps {
		// Apply renames
		if step.Rename != nil {
			for oldName, newName := range step.Rename {
				if v, ok := current[oldName]; ok {
					current[newName] = v
					delete(current, oldName)
				}
			}
		}

		// Add new fields with null defaults
		for _, field := range step.AddedFields {
			if _, ok := current[field]; !ok {
				current[field] = nil
			}
		}

		// Remove old fields
		for _, field := range step.RemovedFields {
			delete(current, field)
		}
	}

	return current
}

// BuildMigrationChain builds the chain from fromVersion to toVersion.
func BuildMigrationChain(
	fromVersion, toVersion int,
	migrations []MigrationStep,
	schemas map[int]*StoredSchema,
) *MigrationChain {
	var steps []MigrationChainStep

	for v := fromVersion + 1; v <= toVersion; v++ {
		var migration *MigrationStep
		for i := range migrations {
			if migrations[i].Version == v {
				migration = &migrations[i]
				break
			}
		}

		prevSchema := schemas[v-1]
		targetSchema := schemas[v]
		if targetSchema == nil {
			continue
		}

		prevFieldNames := make(map[string]bool)
		if prevSchema != nil {
			for _, c := range prevSchema.Columns {
				prevFieldNames[c.Name] = true
			}
		}

		targetFieldNames := make(map[string]bool)
		for _, c := range targetSchema.Columns {
			targetFieldNames[c.Name] = true
		}

		renameOldNames := make(map[string]bool)
		renameNewNames := make(map[string]bool)
		var renames map[string]string
		if migration != nil {
			renames = migration.Rename
			for old, new_ := range renames {
				renameOldNames[old] = true
				renameNewNames[new_] = true
			}
		}

		var addedFields []string
		for name := range targetFieldNames {
			if !prevFieldNames[name] && !renameNewNames[name] {
				addedFields = append(addedFields, name)
			}
		}

		var removedFields []string
		for name := range prevFieldNames {
			if !targetFieldNames[name] && !renameOldNames[name] {
				removedFields = append(removedFields, name)
			}
		}

		steps = append(steps, MigrationChainStep{
			FromVersion:   v - 1,
			ToVersion:     v,
			Rename:        renames,
			AddedFields:   addedFields,
			RemovedFields: removedFields,
			TargetSchema:  targetSchema,
		})
	}

	return &MigrationChain{Steps: steps}
}

// DeserializeWithSchema maps positional field values to a named row using a stored schema.
func DeserializeWithSchema(values []interface{}, stored *StoredSchema) map[string]interface{} {
	row := make(map[string]interface{}, len(stored.Columns))
	for i := 0; i < len(stored.Columns) && i < len(values); i++ {
		row[stored.Columns[i].Name] = values[i]
	}
	return row
}
